File: destination_directory.py
import os


def create_season_path(media_dictionary):
    md = media_dictionary
    if md["Media Type"] == "Movie":
        season_path = ''
    elif md["Media Type"] != "Movie":
        if int(md['Season']) < 10: season_path = f"Season 0{md['Season']}"
        if int(md['Season']) > 9: season_path = f"Season {md['Season']}"
    else:
        print("Season path creation error.")
    return season_path


class DestinationDirectory:

    def __init__(self, base_path, media_dictionary, mode):
        self.destination_directory = ''
        self.base_path = base_path
        self.media_path = ''
        self.season_path = ''
        self.media_dictionary = media_dictionary
        self.media_type = self.media_dictionary["Media Type"]
        self.mode = mode

        # print(media_dictionary)
        if self.base_path:
            self.base_path = self.create_new_base_path()

        self.media_path = self.create_media_path()
        self.season_path = create_season_path(self.media_dictionary)

        self.destination_directory = self.create_destination_directory()


    def create_new_base_path(self):
        if self.media_type == 'TV':
            base_path = 'X:\\TV Shows'
        elif self.media_type == 'Anime':
            base_path = 'X:\\Anime\\Shows'
        elif self.media_type == 'Movie':
            base_path = 'X:\\Movies'
        else:
            print("Error")
        return base_path

    def create_media_path(self):
        md = self.media_dictionary
        media_path = f"{md['Show Name']} [{md['Scene']}][{md['Resolution']}][{md['Source']}][{md['Video Format']}][{md['Audio Format']}]"
        return media_path

    def create_destination_directory(self):
        if self.mode == "Rename":
            destination_directory = self.media_dictionary["Source Directory"]
            print(destination_directory)
            return destination_directory
        elif self.mode == "Hardlink":
            destination_directory = os.path.join(self.base_path, self.media_path, self.season_path)
            print(destination_directory)
            return destination_directory
        elif self.mode == "Preview":
            destination_directory = os.path.join(self.base_path, self.media_path, self.season_path)
            print(destination_directory)
            return destination_directory

File: MediaNamer/DestinationDirectory.cs
using System;
using System.IO;

namespace MediaNamer
{
    public class DestinationDirectoryClass
    {
        public string DestinationDirectory { get; private set; }
        public string BasePath { get; private set; }
        public string MediaPath { get; private set; }
        public string SeasonPath { get; private set; }
        public MediaDictionary MediaDictionary { get; private set; }
        public string MediaType { get; private set; }
        public string Mode { get; private set; }

        public DestinationDirectoryClass(bool basePathFlag, MediaDictionary mediaDictionary, string mode)
        {
            DestinationDirectory = "";
            BasePath = basePathFlag ? "temp" : ""; // base_path true in original, create_new_base_path sets it
            MediaPath = "";
            SeasonPath = "";
            MediaDictionary = mediaDictionary;
            MediaType = MediaDictionary.MediaType;
            Mode = mode;

            if (basePathFlag)
            {
                BasePath = CreateNewBasePath();
            }

            MediaPath = CreateMediaPath();
            SeasonPath = CreateSeasonPath(MediaDictionary);

            if (!string.IsNullOrEmpty(BasePath) && !string.IsNullOrEmpty(MediaDictionary.ShowName))
            {
                string standardTag = $"[{MediaDictionary.Scene}][{MediaDictionary.Resolution}][{MediaDictionary.Source}][{MediaDictionary.VideoFormat}][{MediaDictionary.AudioFormat}]";

                if (Directory.Exists(BasePath))
                {
                    var dirs = Directory.GetDirectories(BasePath, $"{MediaDictionary.ShowName} [*");
                    if (dirs.Length > 0)
                    {
                        string existingDirName = Path.GetFileName(dirs[0]);
                        MediaPath = existingDirName;
                        if (!existingDirName.Contains(standardTag) && MediaDictionary.MediaType != "Movie")
                        {
                            SeasonPath += $" {standardTag}";
                        }
                    }
                }
            }

            DestinationDirectory = CreateDestinationDirectory();
        }

        public static string CreateSeasonPath(MediaDictionary md)
        {
            string seasonPath = "";
            if (md.MediaType == "Movie")
            {
                seasonPath = "";
            }
            else if (md.MediaType != "Movie")
            {
                if (int.TryParse(md.Season, out int seasonInt))
                {
                    if (seasonInt < 10) seasonPath = $"Season 0{seasonInt}";
                    if (seasonInt > 9) seasonPath = $"Season {seasonInt}";
                }
                else
                {
                    Console.WriteLine("Season path creation error: Season is not a valid integer.");
                }
            }
            else
            {
                Console.WriteLine("Season path creation error.");
            }
            return seasonPath;
        }

        public static string GetBasePath(string mediaType)
        {
            if (mediaType == "TV") return @"X:\TV Shows";
            if (mediaType == "Anime") return @"X:\Anime\Shows";
            if (mediaType == "Movie") return @"X:\Movies";
            return "";
        }

        private string CreateNewBasePath()
        {
            string basePath = GetBasePath(MediaType);
            if (string.IsNullOrEmpty(basePath))
            {
                Console.WriteLine("Error");
            }
            return basePath;
        }

        private string CreateMediaPath()
        {
            var md = MediaDictionary;
            string mediaPath = $"{md.ShowName} [{md.Scene}][{md.Resolution}][{md.Source}][{md.VideoFormat}][{md.AudioFormat}]";
            return mediaPath;
        }

        private string CreateDestinationDirectory()
        {
            string destinationDirectory = "";
            if (Mode == "Rename")
            {
                destinationDirectory = MediaDictionary.SourceDirectory;
                Console.WriteLine(destinationDirectory);
                return destinationDirectory;
            }
            else if (Mode == "Hardlink" || Mode == "Preview")
            {
                destinationDirectory = Path.Combine(BasePath, MediaPath, SeasonPath);
                Console.WriteLine(destinationDirectory);
                return destinationDirectory;
            }
            return destinationDirectory;
        }
    }
}
